packages/discovery-provider/src/tasks/entity_manager/entities/email.py

```diff
@@ -143,37 +144,52 @@ def grant_email_access(params: ManageEntityParameters) -> None:
                 f"Missing required fields: {', '.join(missing_fields)}"
             )
 
-        # Verify granter has access
-        granter_access = (
-            params.session.query(EmailAccess)
-            .filter(
-                EmailAccess.email_owner_user_id
-                == params.metadata["email_owner_user_id"],
-                EmailAccess.receiving_user_id == params.metadata["grantor_user_id"],
+        # Validate access_grants structure
+        access_grants = params.metadata["access_grants"]
+        if not isinstance(access_grants, list):
+            raise IndexingValidationError("access_grants must be a list")
+
+        for grant in access_grants:
+            if not all(
+                k in grant
+                for k in ["receiving_user_id", "grantor_user_id", "encrypted_key"]
+            ):
```

Comment:
nit: can do `if not {"receiving_user_id", "grantor_user_id", "encrypted_key"}.issubset(grant):`

Reply:
future consideration, we probably don't need to fail the entire thing if one grant is malformed. 

[PAY-3886]

Reply:
updated here & above usage that i copied